### PR TITLE
Datastore and statement metadata

### LINKED ISSUE #36: Metadata store
KV store for actual metadata objects. Should conform to the [go-datastore](https://github.com/ipfs/go-datastore) interface so IPFS can read from it. Will also serve as the "document store" for the indexer. Can be backed by e.g. LMDB though I suspect @autoencoder has thoughts on this

Closes #33

### PR DESCRIPTION
**Pull request by vyzo**
Closes #36 

- Implements the datastore for statement metadata with rocksdb.
- Provide a `/data/put` and `/data/get` API endpoints for storing/retrieving metadata objects.
- Implement the `/mediachain/node/data` protocol for retrieving statement metadata from remote peers.
- Merge statement metadata along with statements in the merge implementation.

Note:
There is a minor change in the remote query protocol, as the end/error messages where refactored to be reused in the data protocol.

Datastore API details: 
- `/data/put` accepts a batch of DataObjects, which are json objects with a data field containing the data base-64 encoded, and returns a stream of object keys (which are B58-encoded SHA256 multihashes)
- `/data/get/{objectId}` retrieves a data object by its key or 404s.

Example:
```
$ head -c 20 /dev/urandom | base64
FCG389RpiSmWkbK96fd0if4gmhw=
$ curl -s -H "Content-Type: application/json" -d '{"data": "FCG389RpiSmWkbK96fd0if4gmhw="}' http://127.0.0.1:9002/data/put
QmT7KTeJYJ7pnvsUk8g56AqFgp1Cqk91HnuhyhmaYZW4dr
$ curl http://127.0.0.1:9002/data/get/QmT7KTeJYJ7pnvsUk8g56AqFgp1Cqk91HnuhyhmaYZW4dr
{"data":"FCG389RpiSmWkbK96fd0if4gmhw="}
```

Merge API changes:
- `/merge` will now fetch metadata referenced by the statements in the result sets, if it doesn't already have them in the datastore. 
  The API returns two integer values (instead of just one), the count of statements and objects that were merged.
  Missing metadata in the peer are treated as errors and abort the merge.

Example:
```
# the dataset contains a bunch of statements which all reference the same object
$ curl -H "Content-Type: application/text" -d 'SELECT * FROM * LIMIT 10' http://127.0.0.1:9004/merge/QmW5vgizcxphSSV22fKxxNCnv9TbaidYDiBCbHR9bGg2RM
10
1
$ curl -H "Content-Type: application/text" -d 'SELECT * FROM * LIMIT 10' http://127.0.0.1:9004/merge/QmW5vgizcxphSSV22fKxxNCnv9TbaidYDiBCbHR9bGg2RM
0
0
$ curl -H "Content-Type: application/text" -d 'SELECT * FROM * LIMIT 100' http://127.0.0.1:9004/merge/QmW5vgizcxphSSV22fKxxNCnv9TbaidYDiBCbHR9bGg2RM
90
0
```

### CODE REVIEW

File: mcnode/api.go [hunk 1]
```diff
@@ -328,6 +327,81 @@ func (node *Node) httpDelete(w http.ResponseWriter, r *http.Request) {
...
+	for {
+		err = dec.Decode(&dao)
+		switch {
+		case err == io.EOF:
+			break loop
+		case err != nil:
+			apiError(w, http.StatusBadRequest, err)
+			return
+		default:
+			batch = append(batch, dao.Data)
```
Comment (parkan) [inline mcnode/api.go]: so we're building up the batch in memory then flushing to datastore on EOF; 

- can this OOM? what's the batch size limit?
- what are the error semantics when some of the batch fails to insert? looks like any errors from datastore return a 500 but some of the objects may have been committed already -- this may be OK given that inserts are basically idempotent, but I would prefer more granularity

Basically I'm not sure that we really need to batch in memory given that the LSM DB design explicitly does this inside their _memtable_ structure -- we should let them do this for us. This way we can also pipeline decoding and inserts and return a more meaningful success/error stream. Seems like adding backpressure to a channel is pretty straightforward.

File: mcnode/api.go [hunk 2]
```diff
@@ -328,6 +327,81 @@ func (node *Node) httpDelete(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprintln(w, count)
 }
 
+// datastore interface
+type DataObject struct {
+	Data []byte `json:"data"`
```
Comment (parkan) [inline mcnode/api.go]: How do these tags work? `Data` is decoded from `data` field in json, i.e. it's `ABC` in `{ "data": "ABC" }`?

File: mcnode/api.go (hunk 1 above)
Comment (vyzo) [inline mcnode/api.go]: It is very unlikely that we have an OOM from this, the batch has to be very large. There is no batch size limit, but this is a user at fault if he makes a batch so large that it OOMs.

The error semantics are atomic write, either the entire batch is written to the database or an error results.

I think we really want to have a batch, so that we can amortize API call latency.

**Comment by parkan** [comment]
🎆
